lms/djangoapps/courseware/management/commands/migrate_student_module_history.py

```diff
@@ -3,65 +3,118 @@
 
+SQL_MAX_INT = 2**31 - 1
 
 class Command(BaseCommand):
     """
-    Command to migrate all data from StudentModuleHistoryArchive into StudentModuleHistory.
+    Command to migrate data from StudentModuleHistoryArchive into StudentModuleHistory.
+    Works from largest ID to smallest.
     """
     help = dedent(__doc__).strip()
+    option_list = BaseCommand.option_list + (
+        make_option('-i', '--index', type='int', default=0, dest='index',
+            help='chunk index to sync (0-indexed) [default: %default]'),
+        make_option('-n', '--num-chunks', type='int', default=1, dest='num_chunks',
+            help='number of chunks to use [default: %default]'),
+        make_option('-w', '--window', type='int', default=1000, dest='window',
+            help='how many rows to migrate per query [default: %default]'),
+        make_option('-s', '--show-range', action='store_true', default=False, dest='show_range',
+            help="show the range that this command would've run over and the max id in that range")
+    )
 
-    @transaction.commit_manually
-    def handle(self, *args, **options):
+    def handle(self, *arguments, **options):
+        if options['index'] >= options['num_chunks']:
+            self.stdout.write("Index {} is too large for {} chunks\n".format(options['index'], options['num_chunks']))
+            exit(2)
+
+        #Set max and min id from the number of chunks and the selected index
+        chunk_size = SQL_MAX_INT / options['num_chunks']
+        min_id = chunk_size * options['index']
+
+        if options['index'] == options['num_chunks'] - 1:
+            max_id = SQL_MAX_INT
+        else:
+            max_id = chunk_size * options['index'] + chunk_size - 1
+
+        #Start at the max id in the selected range, so the migration is resumable
+        active_range = StudentModuleHistory.objects.filter(id__lt=max_id, id__gte=min_id).order_by('id')
+        if active_range:
+            min_id_found = active_range[0].id
+            self.stdout.write("Found min existent id {} in StudentModuleHistory, resuming from there\n".format(min_id_found))
+        else:
+            #Assume we're starting from the top of the range
+            min_id_found = max_id
+            self.stdout.write("No entries found in StudentModuleHistory in this range, starting at top of range ({})\n".format(min_id_found))
+
+        #Make sure there's entries to migrate in StudentModuleHistoryArchive in this range
         try:
-            max_id = StudentModuleHistory.objects.all().order_by('id')[0].id
-        except IndexError:
-            self.stdout.write("No entries found in StudentModuleHistory, aborting migration.\n")
+            StudentModuleHistoryArchive.objects.filter(id__lt=min_id_found, id__gte=min_id)[0]
+        except:
+            self.stdout.write("No entries found in StudentModuleHistoryArchive in range {}-{}, aborting migration.\n".format(
+                min_id_found, min_id))
+            exit(1)
```

Comment:
It will, but it'll make a lot of noise while it iterates through the range. I figured it'd be better to make it explicit. On the other hand, perhaps exiting with a nonzero status is not appropriate.